### ISSUE TITLE
Performance of the audio processing nodes is sub-par.

### ISSUE THREAD

**Issue by godofecht** [sh-lee97/grafx#1]
For a graph based library, the internal nodes should probably be written in a better language than Python.
As a result the performance isn't comparable to similar solutions in c++ when it could be.

**Comment by sh-lee97** [comment]
Thanks for the suggestion! 

This library and all the processors included were developed for the research of [differentiable digital signal processing (DDSP)](https://arxiv.org/abs/2001.04643). These audio processors (nodes) are specifically designed so that gradient descent (or its variants) can optimize their parameters. To achieve this, it requires (1) some extra steps (e.g., [frequency sampling method](https://dafx2020.mdw.ac.at/proceedings/papers/DAFx2020_paper_7.pdf) for IIR filters) or (2) developing a different filter that might not be that compute-efficient but serves this purpose well (e.g., the provided zero-phase FIR EQ and filtered noise reverb). Also, they are suitable for fast batch processing on GPUs, not CPUs. As you said, some speedups can be achieved by going to C++, but we will try other alternatives first: `torch.compile`, `numba.jit`, and `triton`. Also, our current primary focus is adding more differentiable processors.

That said, with some additions, our library can be used with any processor unless you need the backpropagations of DDSP. For instance, we may want to construct a processing graph with [`Spotify Pedalboard`](https://github.com/spotify/pedalboard); we have done this in several research projects in our lab, e.g., for speech enhancement (data augmentation). Is this the functionality you suggested?
